[MENTORING-T-123] 로또 스프링 구현 과제

Karatuss opened the pull request:
README.md 파일에 진행 상황에 대해 표시해놓았습니다.

### History

24.03.14 1차 PR 리뷰

### TODO

`1순위`
- [x] 나머지 필수 적용 사항들 및 선택 적용 사항들 개발 완료하기
- [x] Entity의 index를 not nullable 하면서도 특별한 의미를 갖는 클래스로 바꿔주기
  - UUID도 고려해보자.
- [x] Index 잡아주기
- [x] DataJpaTest용 [schema, data].sql 설계하기
- [x] RestDocs 테스트로 완성하기
- [x] 동시성 문제 해결하기
  - 예상 시나리오 1: 유저가 createLottos() 서비스에 접근할 때 스레드를 나눠주는 것이 좋아보임.
  - 사용 기능 1: launch()로 따로 결과를 받지 않도록 리펙토링? async()로 기존 response를 받을까?
  - 예상 시나리오 2: 현재 Scheduled로 생성되는 당첨 로또가 컨텍스트 스위칭 비용이 발생하는 기능인지 알아보고 바꿔보기
- [x] docker-compose 작성해서 배포하기

`2순위`
- [x] `LottoNumber` 클래스를 data class -> value class로 바꿔보기
  - AttributeConverter 시도해보기
- [ ] Service 영역에서 Validator들을 컴포넌트로 분리시키기
- [ ] 각 도메인 별로 Exception 만들어주기
- [ ] Scheduled에 다음 라운드 시작되기 전 이전 라운드에 구매된 로또들 다 맞춰보고 유저들에게 결과 push하는 기능
- [ ] 스파이 로거 데코레이팅 ([댓글 링크](https://github.com/BI-kotlin-in-action-2ND/HyeonminShin-LottoSpring/pull/1#discussion_r1526394509))

Karatuss committed:
update: spy.log 제외

Karatuss committed:
build: 필요한 디펜던시 및 RestDoc 설정 추가

Karatuss committed:
docs: 24.03.14 완료 및 진행중 표시 업데이트

Karatuss committed:
chore: application 패키지 메니저 수정

application.yml: 전체 설정
application-database.yml: 데이터베이스 설정

Karatuss committed:
feat: ConfigurationPropertiesScan 애노테이션 추가

Karatuss committed:
feat: 프로젝트 전역 설정 클래스 추가

Karatuss committed:
feat: 프로젝트 전역 예외 처리 관련 클래스 추가

Karatuss committed:
feat: 프로젝트 전역 도메인 클래스 추가

Karatuss committed:
feat: 프로젝트 전역 페이징 Response 클래스 추가

Karatuss committed:
feat: Lotto 컨트롤러 추가

Karatuss committed:
feat: Lotto 관련 도메인 추가

Karatuss committed:
feat: Lotto Dto 추가

Karatuss committed:
feat: Lotto Repository 추가

Karatuss committed:
feat: Lotto Service 추가

Karatuss committed:
feat: User 컨트롤러 추가

Karatuss committed:
feat: User 도메인 추가

Karatuss committed:
feat: User Dto 추가

Karatuss committed:
feat: User Repository 추가

Karatuss committed:
feat: User Service 추가

Karatuss committed:
feat: WinningLotto 컨트롤러 추가

Karatuss committed:
feat: WinningLotto 도메인 추가

Karatuss committed:
feat: WinningLotto Dto 추가

Karatuss committed:
feat: WinningLotto Repository 추가

Karatuss committed:
feat: WinningLotto Service 추가

Karatuss committed:
test: Lotto Repository, Service 테스트 추가

Karatuss committed:
test: User Repository, Service 테스트 추가

Karatuss committed:
test: WinningLotto Repository, Service 테스트 추가

Karatuss committed:
test: User Controller 테스트 draft 추가

Karatuss committed:
test: Kotest용 Transactional 애노테이션 설정 추가

Karatuss committed:
test: 테스트용 애노테이션 및 영속성 메니저 추가

Karatuss committed:
test: 테스트용 Fixtures 추가

- UserFixtures.kt
- LottoFixtures.kt

Karatuss committed:
test: schema-h2.sql draft 추가

konempty commented on .gitignore:
```diff
@@ -38,3 +38,4 @@ out/
 
 ### Kotlin ###
 .kotlin
+/spy.log
```
오호 스프링 자체 JPA 쿼리로거로도 충분할것 같은데 스파이 로거를 쓰시는 이유가 궁금합니다~!

Karatuss replied:
기본 제공되는 JPA 쿼리로거가 가독성이 떨어지는 것 같아 처음에 선택하게 되었습니다. 지금은 기본 설정인데 나중에 [여기 블로그](https://jong-bae.tistory.com/23) 참고해서 디버깅할 때 좀 예쁘게 보려고 합니다 ㅎ

konempty replied:
오호 JPA 기본 로거가 못생겼다는 생각을 못해봤는데 그렇군요 ㅋㅋㅋ

TwoPair replied:
아 제가 용어를 잘못 선택해서 사용했네요. 정확히는 페이지네이션을 지원하는 클래스 중 Page, Slice가 존재하잖아요. 아시다시피 Page는 전체 결과를 가져오고 Slice는 요청한 사이즈만큼 가져오는 클래스인만큼 클래스명 위주의 답변을 작성하다보니 용어의 혼선이 생겼던 것 같습니다 :)

Karatuss replied:
아 제가 용어를 잘못 선택해서 사용했네요. 정확히는 페이지네이션을 지원하는 클래스 중 Page, Slice가 존재하잖아요. 아시다시피 Page는 전체 결과를 가져오고 Slice는 요청한 사이즈만큼 가져오는 클래스인만큼 클래스명 위주의 답변을 작성하다보니 용어의 혼선이 생겼던 것 같습니다 :)

konempty commented on build.gradle.kts:
```diff
@@ -1,51 +1,107 @@
-	kotlin("plugin.jpa") version "1.9.22"
+    id("org.springframework.boot") version "3.2.3"
+    id("io.spring.dependency-management") version "1.1.4"
+    id("org.asciidoctor.jvm.convert") version "3.3.2"
```
스웨거 VS 아스키독 중에 아스키독을 선택하게된 이유가 궁금합니다~!

Karatuss replied:
참고한 우아한 기술 블로그: [Spring Rest Docs 적용](https://techblog.woowahan.com/2597/)

고민 많이 하다가 위의 블로그에서 정리한 글을 보았을 때 결심을 했었습니다. 그래서 정리한 **아스키독 사용 이유**는 다음과 같습니다.

1. 제품 코드에 영향이 없다.
2. 프로덕션 코드에 영향이 없다.
3. 즉 코드가 깔끔해진다.
4. 실무에서 많이 사용하는 스펙이라 알고 있는데 미리 사용해보고 싶다.
5. 기왕 테스트 작성할거 성공할 때 같이 반영되면 좋을 것이다.

혹여나 이해되지 않으시거나 제가 잘못 생각한 부분이 있다면 피드백 해주시면 감사하겠습니다 😄

konempty replied:
네넵 요건 갠취의 영역이라 이유만 있으면 OK입니다~!

Karatuss replied:
> id가 not null로 바뀐건 좋은데 id 생성 전략까지 시퀀스로 바뀐이유가 있을까요?? 오라클 디비 아니면 보통 identity로 충분하고 시퀀스를 쓰는게 독특해서 질문드려봅니다~!

아뇨 그런 것까진 잘 모르고 사용했습니다. 그런 이유가 있었군요. 인터넷 글들을 참고하면서 쓰다보니 맞는 환경인지 고려 안하면서 가져온게 이렇게 된 것 같습니다! 추후 MySql, 또는 Postgre 쓸 예정인데 해당 시퀀스는 다시 없애보겠습니다.

> 그리고 한가지 추가 질문드리고 싶은게 있는데 TODO로 id컬럼을 UUID로 정의해볼생각도 있다고 하셨는데 UUID가 string형태로 저장되다보니 Long형인 id에 비해 성능이 일부 떨어지는데 요건 고려가 된 설계일까요??

성능이 떨어지는 건 알고 있었습니다만, 사람들이 보통의 경우 **대량의 로또를 한 번에 구입하는 것**을 생각하면 Long의 범위를 넘어선, 어쩌면 ULong을 넘어선 범위까지도 갈 수 있겠다 싶었습니다. 그럼에도 UUID를 사용했을 땐 다량의 로또 생성시 성능 이슈가 발생할 수 있다고는 생각합니다.
DB 증설할 때 네이밍을 다르게 가져가는 규칙같은게 있다면 편하게 Long(ULong) 써도 될 것 같습니다만, 아직 제가 그런 솔루션에 대해 잘 모르는지라 아시는 해결책이 있으시다면 알려주시면 감사하겠습니다!

konempty commented on build.gradle.kts:
```diff
@@ -1,51 +1,107 @@
-		jvmTarget = "21"
-	}
+    compileOnly("org.projectlombok:lombok")
+    annotationProcessor("org.projectlombok:lombok")
```
오 코틀린을 쓰는데 롬복의 도움이 필요한 코드가 있엇나요??

Karatuss replied:
아.. 빌드 긁어온거 처리 못한 녀석이 있었군요.. 제거했습니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/controller/LottoController.kt:
```diff
@@ -0,0 +1,54 @@
+        @Positive(message = "조회할 개수는 양수여야 합니다.")
+        @RequestParam("size", defaultValue = "1", required = false)
+        size: Int?,
+    ): ResponseEntity<SliceResponse<LottoResponse>> {
```
오호 Pagination대신에 Slicing을 선택한 이유가 궁금합니다~!

Karatuss replied:
아무래도 조회하는 로또의 개수는 유저가 구매한 로또를 기반해서 가져오는 경우가 흔할텐데, 한 사람의 유저가 대량의 로또를 구매한 경우를 생각해봤습니다.

Pagination의 경우 아무래도 **전체** 로또를 가져오려고 하다보니 브라우저에 표시할 때도 부담스럽고 다중 요청시 서버의 부하 처리도 고민이 될 것입니다. 따라서 일단 몇 개를 보여준 뒤, 추가로 요청하면 또 조회해서 응답하는 식으로 하는 것이 **성능상 이점**이 있을 것이라 생각해 Slicing을 선택했습니다.

Karatuss replied:
1. 네이밍은 LottoNumber**s**, 프로퍼티명은 number**s**로 바꾸겠습니다.
2. `java.lang.VerifyError: Bad local variable type` 때문입니다.
```kotlin
Type top (current frame, locals[0]) is not assignable to reference type"]
```
위와 같은 에러가 발생해서 원인을 추적해봤습니다. 디컴파일을 통해 확인해본 결과, Collection<LottoNumber>라고 쓰는 경우, WrapperClass인 Collection(reference type) 안에 LottoNumber(primitive로 변환됨)을 넣으려고 해서 생기는 문제점인 것을 발견했습니다. 
이를 해결하기 위해 IntArray 같은 자료형을 쓸 수도 있었겠지만, **코드 복잡도 증가 및 추후 유지보수 측면**에서도 어려울 것으로 판단했습니다. 따라서 전체 콜렉션을 저장하는 일급 객체로서 사용하자는 생각을 했고요, 기존 Lotto에 존재하던 로또 번호 개수 const 변수도 아예 안으로 편입시켜 POJO에 더 가깝게 설계할 수 있었습니다.

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/domain/Lotto.kt:
```diff
@@ -0,0 +1,66 @@
+class Lotto(
+    @Column(nullable = false, updatable = false)
+    val round: Int,
+    numbers: SortedSet<LottoNumber>,
```
여기서 받고 아래에서 numbers라는 프로퍼티에 재할당 하는이유가 있을까요??
그리고 로직상에서는 numbers가 변경될일이 없는것 같은데 var로 선언된 이유도 궁금합니다~!

Karatuss replied:
코드 수정 중에 놓쳤습니다 ㅎㅎ
저도 val이 맞다고 생각해서 고쳤어요!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/domain/Lotto.kt:
```diff
@@ -0,0 +1,66 @@
+    val user: User,
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "lotto_id")
```
이건 컨벤션의 영역이라 리뷰드리긴 쪼금 애매하지만
보통 FK명명 규칙이 ~~_id라서 해당 엔티티의 PK컬럼명을 lotto_id라고 지으면
다른사람이 보기엔 FK라고 오해할것 같아요
저도 순간 오해했구요 ㅋㅋㅋㅋ...

Karatuss replied:
ㅋㅋㅋ그럴 것 같습니다. 이것도 지웠습니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/domain/Lotto.kt:
```diff
@@ -0,0 +1,66 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "lotto_id")
+    val id: Long? = null,
```
id 컬럼이 nullable로 선언된 이유가 궁금합니다!
아마 DB에 컬럼이 생성되기전 ID를 알수없어서 그랬을것 같은데
다른 방법은 없었을까요??

Karatuss replied:
요거 같은 경우에는 저도 고민이 많았고 여러 사람들의 블로그와 문서들을 뒤져본 뒤 설계한 엔티티 형태입니다.

### Reference
- [Spring 공식 문서](https://spring.io/guides/tutorials/spring-boot-kotlin)
> [!NOTE]
> Entity 선언 예제 발췌 일부
```kotlin
@Entity
class Article(
    var title: String,
    var headline: String,
    var content: String,
    @ManyToOne var author: User,
    var slug: String = title.toSlug(),
    var addedAt: LocalDateTime = LocalDateTime.now(),
    @Id @GeneratedValue var id: Long? = null)
``` 
- [관련 블로그](https://multifrontgarden.tistory.com/304)

해당 형태에 대해서는 저도 의문이 많았는데요 결국 다음의 이유로 위와 같은 형태가 된 것으로 봅니다.

### 이유

1. 생성자 내부 존재 이유: 생성자 내부가 아닌 예를 들어 class body에 넣게 된다면, 단위 테스트 시 영속화되지 않으면서 id를 설정할 때 외부에서 넣어줄 수 없기 때문에 테스트가 힘들어집니다.
2. 생성자 최하단 선언 이유: 최상단에 넣어버리면 코틀린 특성상 기본 생성자 위치가 곧 생성자 파라미터 위치가 되어버리므로 `Article(null, "testTitle", ...)`과 같은 형태가 되어버리므로 반복적으로 null을 넣어주어야 하므로 불편함이 생깁니다.

따라서 공식에서도 다음과 같은 가이드라인을 제시한 것이 아닌가 싶고 저도 이에 맞춰 구현해보았습니다.

Karatuss replied:
하지만 이 문제는 추후 어떤 식으로든 엔티티 생성 시점에 id값이 들어가야 하는 것이 동일성 검증 등에 사용되므로 **NOT Nullable 한 것이 맞다고 생각**합니다.

그래서 리펙토링할 때 단순히 Long을 사용하는 것이 아니라 **UUID 또는 Id를 위한 식별자 클래스**를 만들어보려 합니다!
todo에 추가했습니다.

Karatuss replied:
현재는 `Long?` -> `Long` 으로 바꾸고 채번 전략을 시퀀스로 잡아주었습니다.
추후에 UUID로도 바꿔볼까 생각중입니다.

Karatuss replied:
로그 시스템 중앙 적체 방식은 잘 몰랐는데 나중에 적용해볼께요!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/domain/Lotto.kt:
```diff
@@ -0,0 +1,66 @@
+
+    fun match(other: Lotto): List<Int> {
+        val intersect = this.numbers.intersect(other.numbers)
+        return intersect.map { it.number }.toList()
```
.map() 의 반환형이 List<Int>일텐데 toList()를 해주는 이유가 궁금합니다~!

Karatuss replied:
실수네요ㅠㅠ toList() 삭제했습니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/dto/LottoSimpleResponse.kt:
```diff
@@ -0,0 +1,19 @@
+            userId: Long,
+            round: Int,
+            size: Int,
+        ) = LottoSimpleResponse(
```
오잉 ㅋㅋㅋㅋ 이렇게되면 정적 팩토리 메소드가 의미가 있나요??

Karatuss replied:
개발 중에 남은 더미 데이터입니다ㅠㅠ 제거했습니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/domain/Lotto.kt:
```diff
@@ -0,0 +1,66 @@
+// TODO: 최적화 솔루션) numbers를 bitset으로 다루기!!
+
+@Entity
+class Lotto(
```
지금 보니까 DDL로는 반영하셨을지는 모르겠지만 요 엔티티 인덱스가 안잡혀있네요 ㅋㅋㅋ..
명시적으로 어떤 인덱스가 잡혀있는지 적어두면 협업에 도움이될것 같아요~!

Karatuss replied:
적용 완료했습니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/domain/LottoNumber.kt:
```diff
@@ -0,0 +1,52 @@
+// TODO: insert 쿼리 너무 많이 나감. 그냥 콤마로 숫자 구분하는 문자열 파서 + 검증기로 만들자.
+@Embeddable
+@NoArgsConstructor
+data class LottoNumber(val number: Int) : Comparable<LottoNumber> {
```
value class로 만드는건 어떨까요??

Karatuss replied:
중간에 equals() 필요해서 잠깐 넣었는데 수정하는 과정에서 더는 필요없어져서 다시 value로 고쳐보겠습니다!

Karatuss replied:
```kotlin
org.springframework.orm.jpa.JpaSystemException: Unable to locate constructor for embeddable 'com.hm.hyeonminshinlottospring.domain.lotto.domain.LottoNumber'
``` 
적용해보니 테스트할 때 위와 같은 에러가 뜨더군요. 구글링도 결과 \@Embeddable 애노테이션이 붙는 클래스의 경우 public constructor를 호출하는 로직이 존재했습니다. value class LottoNumber를 디컴파일 해보니 초기화 로직이 private constructor와 static class initilizer로 이루어져 있어 맞지 않다고 되더라고요.

그래서 생각한게 일단 지금은 data class로 유지하되 이후 패치에서 다른 형태로 LottoNumber를 바꿔볼 생각을 하고 있습니다.

Karatuss replied:
value class LottoNumber로 리펙토링 했습니다!

Karatuss replied:
schema.sql, data.sql 작업을 못해서 그냥 통합 테스트로 때웠습니다 ㅎㅎ..
해당 파일들 생성하고 바꾸고 다시 테스트해 볼 예정입니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/dto/LottoCreateRequest.kt:
```diff
@@ -0,0 +1,39 @@
+        round: Int,
+        user: User,
+        numbers: Collection<Int>,
+    ) = Lotto(
```
엇 코드가 좀 이상하네요..?
해당 RequestDTO를 엔티티로 바꾸는 메소드가 아니라
해당 메소드에 값을 넣어줘서 해당 값으로 메소드를 만드는군요..??
이름(해당 DTO가 엔티티가 될것이라 기대)과 
동작(DTO의 값을 활용하지 않고 전부 외부로 부터 값을 받아 그걸 엔티티로 만듦)이 일치하지 않네요 ㅠㅠ

Karatuss replied:
리펙토링을 통해 이름과 동작이 일치하게 수정했습니다.
여전히 이상하다면 피드백 주시면 감사하겠습니다!

Karatuss replied:
당연히 케바케라고 생각은 드는데요, 저 같은 경우는 테스트 유저의 unique 속성이 자꾸 외래키 에러 떠서 고치려고 적용해본 것입니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/service/LottoService.kt:
```diff
@@ -0,0 +1,131 @@
+    ): SliceResponse<LottoResponse> {
+        validateUserExistence(userId)
+        validateUserIsNotAdmin(userId)
+        validateSizeAndRoundIsPositive(size, round)
```
요건 컨벤션의 영역이지만 제가 선호하는 방식을 알려드리자면
요런 validation역할은 서비스의 역할이 아니라고 여기고 OOOValidatior라는 별도의 컴포넌트로 몰아두는걸 선호합니다~!
요건 갠취라 적용 안하셔도 OK입니다!

Karatuss replied:
우선순위를 낮춘 todo로 등록해서 추후 적용해보겠습니다.

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/service/LottoService.kt:
```diff
@@ -0,0 +1,131 @@
+                            ),
+                        )
+                    }.onSuccess { successCount += 1 }
+                }
```
헉 ㅋㅋㅋ save를 반복문 돌려서 저장해야하는 니즈가 있다면
saveAll메소드로 한번에 저장하는게 효율적입니다~!
요 아래 블로그를 읽어보시면 도움이 될것 같습니다~!
https://maivve.tistory.com/342

Karatuss replied:
적용해보겠습니다 감사합니다!!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/service/LottoService.kt:
```diff
@@ -0,0 +1,131 @@
+        when (request.mode) {
+            GenerateMode.RANDOM ->
+                repeat(request.generateCount) {
+                    runCatching {
```
save로직이 실패할 염두를 하고 runCatching로 감싼것 같은데 요건 좀 의외네요 ㅋㅋㅋ
보통 save로직이 실패하는건 2가지 이유죠
1. 네트웍이슈 혹은 하드웨어 페일로 DB와 커넥션이 중간에 끊김
2. 제약조건 통과 실패

하지만 1번의 경우 한번 발생하면 수초 ~ 수분, 심지어 수시간동안 실패 할거라 여기에 runCatching이 있는 의미가 퇴색될것 같고
오히려 완전 실패를 시키는게 안전하고
2번은 어플리케이션에서 검증을 하는것에 문제가 있어서 해당 이슈가 발생한것이기에 로직을 탄탄히 해야하는 이유가 되지
에러를 처리할 이유는 아닐것 같아요!

혹시  어플리케이션이 해야할 제약조건을 DB에게 위임해서 검증하려는 의도였다면 DB의 제약조건 검증 비용은
어플리케이션에서 하는것 보다 훨씬 비싼 실패연산이기때문에 비추천드려요 ㅋㅋㅋ

Karatuss replied:
로직에서 제외했습니다, 감사합니다!
로컬 프로그래밍 단위에서의 에러 수준만 고려하다보니 실무 환경과 연관짓는 발상은 하지 못했네요.
덕분에 좀 더 고급화된 프로그래밍이 가능할 것 같습니다 ㅎㅎ

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/controller/LottoController.kt:
```diff
@@ -0,0 +1,54 @@
+        round: Int,
+        @Positive(message = "조회할 개수는 양수여야 합니다.")
+        @RequestParam("size", defaultValue = "1", required = false)
+        size: Int?,
```
이 값이 null일수가 있나요..??

Karatuss replied:
로직상 size 파라미터 안주면 1개 리턴하게 해야겠다고 생각했는데, 다시 생각해보니 구체적인 개수를 달라고 하는게 더 안정적인 api일 것 같아 null이 아니도록 고쳤습니다!

Karatuss replied:
현재는 Pageable 적용으로 삭제처리 했습니다.

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/user/controller/UserController.kt:
```diff
@@ -0,0 +1,70 @@
+        userCreateRequest: UserCreateRequest,
+    ): ResponseEntity<UserResponse> {
+        val response = userService.createUser(userCreateRequest)
+        return ResponseEntity.ok(response)
```
해당 케이스에는 HTTP Status 코드의 의미에 맞게
200(OK) 대신 201(created)를 내려주는게 어떨까요??

Karatuss replied:
수정했습니다 :)

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/user/controller/UserController.kt:
```diff
@@ -0,0 +1,70 @@
+        return ResponseEntity.ok(response)
+    }
+
+    @PutMapping("/{userId}/addMoney")
```
유저의 정보중에 일부만 수정하는거니까 patch가 맞지않을까요??

Karatuss replied:
addMoney, withdrawMoney 둘 다 PatchMapping으로 수정했습니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/user/domain/User.kt:
```diff
@@ -0,0 +1,42 @@
+
+    fun withdrawMoney(withdrawMoney: Int): Int {
+        require(withdrawMoney > 0) { "출금할 값을 양수로 입력해주세요." }
+        money -= withdrawMoney
```
가지고 있는 금액보다 더 꺼내는건 확인 안해도 될까요?? ㅋㅋㅋㅋ

Karatuss replied:
하마터면 대출 시스템을 구현할 뻔 했군요.. 조건문 추가했습니다 ㅎㅎ

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/user/repository/UserRepository.kt:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Repository
+
+fun UserRepository.existsByUserId(userId: Long): Boolean = existsById(userId)
```
헉 ㅋㅋㅋ 이건 확장함수로 만들 이유가 있었을까요?? ㅋㅋㅋㅋ

Karatuss replied:
뭔가 userId를 함수명에 포함시키는 편이 보기 편할 것이라 생각했는데, 뒤에 아무런 조치없이 단순히 이름만 가져올거라면 abundant한 구현이겠죵..
삭제처리하고 리펙토링 했습니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/user/service/UserService.kt:
```diff
@@ -0,0 +1,77 @@
+    @Transactional(readOnly = true)
+    fun getUserInformation(userId: Long): UserResponse {
+        validateUserIdExistence(userId)
+        val user = userRepository.getByUserId(userId)
```
어차피 유저가 없으면 getByUserId에서 에러가 발생할텐데 validateUserIdExistence로 검증하는 의미가 있을까요?

Karatuss replied:
그렇네요. 삭제했습니다!
마찬가지로 중복되는 예외 검사 루틴을 전부 삭제 했습니다.

TwoPair replied:
마찬가지로 round, size가 양수인지 검사하는 `validateSizeAndRoundIsPositive()` 함수 또한 컨틀롤러 단에서 걸러지므로 삭제처리 했습니다.
아니면 이건 그대로 둬야할까요?

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/winninglotto/cotroller/WinningLottoController.kt:
```diff
@@ -0,0 +1,60 @@
+class WinningLottoController(
+    private val winningLottoService: WinningLottoService,
+) {
+    @Scheduled(fixedRate = PER_ROUND_TIME)
```
서버가 켜진 뒤로 부터 PER_ROUND_TIME이라는 시간이 지나면 실행되는것 같은데
로또라면 서버의 재기동 시간과는 상관없이 정해진 시간에 당첨번호가 나와야 하지 않을까요?? ㅋㅋㅋㅋ

konempty replied:
앗 요건 갠취이긴한데 스케쥴러가 컨트롤러단에 있는것도 조금 불편하네요 ㅋㅋㅋㅋ..

Karatuss replied:
cron으로 바꾸고 Service 단으로 옮겼습니다 ㅋㅋㅋ

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/winninglotto/domain/WinningLottoInformation.kt:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.stereotype.Component
+
+@Component
+class WinningLottoInformation(round: Int = 1) {
```
여기서 round는 받아놓고 사용을 안하는것 같은데 의도된건가요? ㅋㅋㅋ

Karatuss replied:
아뇽 더미 데이터입니다 ㅎ
해당 부분 삭제 처리 했습니다.

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/service/LottoService.kt:
```diff
@@ -0,0 +1,131 @@
+        request: LottoCreateRequest,
+    ): LottoSimpleResponse {
+        val user = userRepository.getByUserId(userId)
+        val round = request.round
```
사용자가 요청을하면 현재 라운드 뿐만 아니라 과거 라운드의 로또도 살수 있는건가요..??
그렇게 하면 과거의 우승 로또번호를 알수 있어서 무조건 맞출수 있지않을까요? ㅋㅋㅋ

Karatuss replied:
수정했습니다 ㅋㅋㅋㅋ

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/global/exception/GlobalExceptionHandler.kt:
```diff
@@ -0,0 +1,140 @@
+        status: HttpStatusCode,
+        request: WebRequest,
+    ): ResponseEntity<Any>? {
+        logger.error("[MethodArgumentNotValidException] ${ex.messages()}")
```
아하 ㅋㅋ... 요렇게 로그를 남기면 나중에 로그를 보고 해당 시점에 어떤 에러가 발생하는지는 알수 있지만
왜 발생했는지 원인 파악은 할수 없어요 
좀더 자세한 로깅을 남겨보는건 어떨까요??
아니면 혹시 별도의 에러 트래킹 모듈 적용을 고려한 로깅일까요??

Karatuss replied:
음 일단 자세히 남겨보라는 말씀에 고치긴 했는데 어떤지 한 번 봐주시겠어요?ㅎㅎ

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/global/support/doamin/NoArgsConstructor.kt:
```diff
@@ -0,0 +1,5 @@
+
+@Target(AnnotationTarget.CLASS)
+@Retention(AnnotationRetention.RUNTIME)
+annotation class NoArgsConstructor
```
오잉 ㅋㅋㅋ 요건 뭘위한 어노테이션일까요??

Karatuss replied:
[여기 코드 부분](https://github.com/BI-kotlin-in-action-2ND/HyeonminShin-LottoSpring/pull/1/files#diff-c0dfa6bc7a8685217f70a860145fbdf416d449eaff052fa28352c5cec1a98c06R19-R21)에 넣을 거였는데 깜빡하고 못 넣었습니다 ㅎㅎ
다음 리뷰 요청 때 반영사항 확인해주시면 감사하겠습니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/global/support/dto/SliceResponse.kt:
```diff
@@ -0,0 +1,18 @@
+    open val hasNext: Boolean,
+    open val content: List<T>,
+) {
+    constructor(size: Int, content: List<T>) : this(
```
으흠 content의 타입이 Slice\<T\>라면 Size를 받지 않아도 괜찮지 않았을까요??
그렇다면 당연히 레포지토리에서부터 반환형이 Slice가 되어야겠죠?

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/global/support/dto/SliceResponse.kt:
```diff
@@ -0,0 +1,18 @@
+
+    constructor(pageable: Pageable, content: List<T>) : this(
+        pageable.pageSize,
+        content,
```
여기도 마찬가지구요

konempty commented on src/main/resources/application-database.yml:
```diff
@@ -0,0 +1,56 @@
+spring:
+  jpa:
+    open-in-view: false
```
오호.. 요걸 끄신 이유가 궁금합니다~!

Karatuss replied:
이건 그냥 클라를 따로 만들걸로 보고 설정해줬습니다. 그리고 db 성능 같은 것도 따지면 좋을까 싶어서요.
혹시 제가 잘못 생각하고 있는 것이 있다면 알려주시면 감사하겠습니다!

konempty replied:
아하 ㅋㅋㅋ 네넵 저게 어떤 설정인지 인지하고 끄신건지 궁금했습니다~!
저도 끄는걸 더 선호해요

konempty commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,11 @@
+  error:
+    whitelabel:
+      enabled: false
+  shutdown: graceful
```
엇 graceful shutdown을 위한 타임아웃 설정같은것도 같이 잡아두시면 좋을것 같아요..!

Karatuss replied:
3초로 타임아웃 잡아뒀습니다

konempty reviewed (changes requested):
어후 ㅋㅋㅋ 고생하셨습니다~!
제가 원래 한번에 리뷰를 20개 이상 안주는데 코드에 너무 집중한 나머지
리뷰를 한번에 너무 많이 드리네요 ㅋㅋㅋ;;;

학생치곤 상당이 구조있고 잘짠 프로젝트 같아서 정말 마음에 들고
코틀린 문법 사용 능력은 제가 손댈게 없을정도로 완벽에가깝네요 ㅋㅋㅋ

일단 이번 리뷰는 여기까지만 드리고 수정후 다시 리뷰 요청 주시면 추가 리뷰 드리겠습니다~!

Karatuss committed:
chore: .editorconfig 추가

Karatuss committed:
build: 빌드 스크립트 리펙토링

Karatuss committed:
docs: adoc 임시 작성

Karatuss committed:
feat: LottoNumber 컨버터 추가

Karatuss committed:
refactor: 변수 위치 변경 반영

Karatuss committed:
feat: 유저 인덱스 추가 및 로또 번호의 value class 반영

Karatuss committed:
feat: data class -> value class

POJO에 맞도록 설계 및 최적화 완료

Karatuss committed:
feat: lotto dto 내부 구조 변경 및 추가

삭제
- LottoResponse.kt
- LottoSimpleResponse.kt

수정
- LottoCreateRequest.kt

생성
- LottoCreateResponse.kt
- SliceLottoNumberResponse.kt

Karatuss committed:
feat: 불필요한 네이밍 변경 및 반환형 변경에 따른 함수명 변경

Karatuss committed:
feat: LottoNumber 변경에 따른 리턴 타입 변경

Karatuss committed:
refactor: 리턴 타입 변경으로 인한 조정

Karatuss committed:
refactor: 서비스 리턴 타입 반영

Karatuss committed:
feat: 테이블 시퀀스 지정 및 withdraw 조건 추가

Karatuss committed:
refactor: 펙토리 메서드들의 파라미터 조정

Karatuss committed:
refactor: 불필요한 레퍼 함수 제거

Karatuss committed:
refactor: 레포지토리 변경에 따른 변경, 불필요한 validator 삭제

Karatuss committed:
refactor: URI 조정

Karatuss committed:
refactor: URI 및 리턴 타입 조정

Karatuss committed:
refactor: 더미 주 생성자 제거

Karatuss committed:
feat: lotto 인덱스, id 시퀀스 추가

Karatuss committed:
feat: 필요한 response 타입 수정 및 추가

Karatuss committed:
refactor: 불필요한 레퍼 함수 제거

Karatuss committed:
refactor: dto 변경 사항 반영

Karatuss committed:
config: JacksonObjectMapper 위한 설정 추가

Karatuss committed:
config: 부 생성자 파라미터 변경

현재 해당 DTO를 사용하지는 않지만, 추후 대량의 데이터 반환 기능 생길 시 사용 예정

Karatuss committed:
refactor: 좀 더 자세한 MethodArgumentNotValidException 에러 메시지

Karatuss committed:
delete: 예전 버전 SliceResponse 삭제

Karatuss committed:
chore: application 설정 파일 수정 및 추가

Karatuss committed:
test: Lotto 관련 테스트 추가

Karatuss committed:
test: User 관련 테스트 추가

Karatuss committed:
test: WinningLotto 관련 테스트 추가

Karatuss committed:
test: boolean 컨버터 추가

Karatuss committed:
test: controller 및 rest docs 테스트 헬퍼 추가

Karatuss committed:
test: 통합 테스트 애노테이션 추가

Karatuss committed:
test: 테스트용 fixtures 수정 및 추가

Karatuss committed:
test: rest docs 용 스니펫 추가

Karatuss commented:
추후 개발 과정들은 위쪽 todo에 등록해놓았습니다. 꾸준히 업데이트 할 예정입니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/controller/LottoController.kt:
```diff
@@ -0,0 +1,53 @@
+        round: Int,
+        @PageableDefault(page = 0, size = 10)
+        pageable: Pageable,
+    ): ResponseEntity<SliceLottoNumberResponse> {
```
페이지네이션을 사용하지 않고 Slice를 택한이유에서
> Pagination의 경우 아무래도 전체 로또를 가져오려고 하다보니 

라고 적어주셨는데 전체를 가져오지 않으려고 페이지네이션을 쓰는게 아닐까요..??
답변이 이해가 잘 안되서 추가 질문드립니다~!

Karatuss replied:
아 제가 용어를 잘못 선택해서 사용했네요. 정확히는 페이지네이션을 지원하는 클래스 중 Page, Slice가 존재하잖아요. 아시다시피 Page는 전체 결과를 가져오고 Slice는 요청한 사이즈만큼 가져오는 클래스인만큼 클래스명 위주의 답변을 작성하다보니 용어의 혼선이 생겼던 것 같습니다 :)

konempty replied:
으흠... 더 이해가 안되는데 Page\<T>가 Slice\<T>를 상속 받는다는거 알고계신가요??
<img width="608" alt="스크린샷 2024-04-06 오후 3 37 34" src="https://github.com/BI-kotlin-in-action-2ND/HyeonminShin-LottoSpring/assets/44064530/3dfa6f6c-2c95-4bd6-9469-db64782d5736">

그래서 Slice의 동작와 Page의 동작은 비슷한 부분이 많고 둘다 DB에서 일부데이터만 가져옵니다!

혹시 아래의 이슈를 의미하신거였다면 Slice도 동일한 이슈를 안고있습니다 ㅠㅠ
https://tecoble.techcourse.co.kr/post/2020-10-21-jpa-fetch-join-paging/

Karatuss replied:
상속 받는다는건 알고 있었는데 제가 정확한 용법을 몰라서 그랬습니다 ㅎㅎ
앞으론 쿼리에 따라 적절한 페이지네이션을 적용하도록 노력해야겠네요!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/domain/Lotto.kt:
```diff
@@ -0,0 +1,63 @@
+    @Id
+    @Column(columnDefinition = "NUMERIC(19, 0)")
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "LOTTO_SEQ_GENERATOR")
+    val id: Long = 0L,
```
id가 not null로 바뀐건 좋은데 id 생성 전략까지 시퀀스로 바뀐이유가 있을까요??
오라클 디비 아니면 보통 identity로 충분하고 시퀀스를 쓰는게 독특해서 질문드려봅니다~!

konempty replied:
그리고 한가지 추가 질문드리고 싶은게 있는데 TODO로 id컬럼을 UUID로 정의해볼생각도 있다고 하셨는데
UUID가 string형태로 저장되다보니 Long형인 id에 비해 성능이 일부 떨어지는데 요건 고려가 된 설계일까요??

Karatuss replied:
> id가 not null로 바뀐건 좋은데 id 생성 전략까지 시퀀스로 바뀐이유가 있을까요?? 오라클 디비 아니면 보통 identity로 충분하고 시퀀스를 쓰는게 독특해서 질문드려봅니다~!

인터넷에 있는 소스를 참고해서 사용한다는 것이 DB에 맞게 설계하지 못한것 같습니다. IDENTITY로 다시 바꾸겠습니다!

> 그리고 한가지 추가 질문드리고 싶은게 있는데 TODO로 id컬럼을 UUID로 정의해볼생각도 있다고 하셨는데 UUID가 string형태로 저장되다보니 Long형인 id에 비해 성능이 일부 떨어지는데 요건 고려가 된 설계일까요??

알고는 있었지만, 보통 다량의 로또를 한 번에 구매하는 경우가 많으니 이 데이터들이 쌓이면 자연스레 Long의 범위를 넘어설 것이라 생각됩니다. 따라서 고려해본 것이 UUID였으나 이 또한 생성시 키 생성 성능에 지대한 영향을 끼칠 것 같다고 생각합니다.
혹시 이에 대한 해결책에 대해 아시는 것이 있나요?

konempty replied:
Long형의 범위가 9,223,372,036,854,775,807 인데 요걸 넘어선다는걸 고려해본 설계는 해본적이 없어서 잘 모르겠네요 ㅋㅋㅋ...
이론상 전세계 80억 인구가 각자 10억개의 로또를 구매해야 넘어설수 있는 구조라고 생각이 드는데 오버엔지니어링일것 같아요..!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/domain/LottoNumber.kt:
```diff
@@ -0,0 +1,35 @@
+import java.util.SortedSet
+
+@JvmInline
+value class LottoNumber private constructor(private val number: SortedSet<Int>) {
```
오호... 이름은 로또번호 하나를 저장할것 처럼 생겼는데 실제로는 로또번호 6개를 다 가지는 클래스가 되어버렸군요..!
혹시 어떤 어려움이 있었길래 LottoNumber가 숫자 하나 받던 클래스에서 다 받는 클래스가 된건지 알수있을까요??

Karatuss replied:
1. LottoNumbers와 numbers로 고치겠습니다.
2. 아래와 같은 에러 발생 때문에 고쳤습니다.
```kotlin
java.lang.Verifyerror: bad local variable type: 
Type top (current frame, locals[0]) is not assignable to reference type
```
처음대로 Int 하나만을 가질 때는 위와 같은 에러가 발생해서 추적해봤습니다. 디컴파일 해보니 Collection과 같은 레퍼 타입 안에 Int로 넣어주는 형태 때문에 에러를 발생시키는 것이더라고요. 그래서 아예 전체를 다 저장하도록 바꿔준 뒤, POJO에 맞도록 Lotto 안에 있던 const 변수인 전체 개수도 안으로 가져와서 만들어줬습니다.

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/domain/converter/LottoNumberAttributeConverter.kt:
```diff
@@ -0,0 +1,16 @@
+@Converter
+class LottoNumberAttributeConverter : AttributeConverter<SortedSet<Int>, String> {
+    override fun convertToDatabaseColumn(entityValue: SortedSet<Int>?): String? {
+        return entityValue?.joinToString(",")
```
👍 💯

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/dto/LottoCreateRequest.kt:
```diff
@@ -0,0 +1,62 @@
+        user: User,
+        round: Int,
+        randomLottoNumbersGenerator: RandomLottoNumbersGenerator,
+    ) = createLottosWithMode(user, round, randomLottoNumbersGenerator)
```
이렇게 쓰이는 메소드라면 굳이 멤버 메소드가 아닌 정적 팩토리 메소드여도 괜찮지 않았을까요??
해당 RequestDTO의 값을 전혀 참조하는것 같지 않아 조언드려봅니다~!

Karatuss replied:
앗 미처 끝까지 확인 못한 부분이군요. 수정하겠습니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/domain/Lotto.kt:
```diff
@@ -0,0 +1,63 @@
+    @JoinColumn(name = "user_id", updatable = false)
+    val user: User,
+    @Id
+    @Column(columnDefinition = "NUMERIC(19, 0)")
```
h2 DB면 Long에 대응하는 기본 컬럼 타입이 NUMERIC이 아닐텐데 요걸 쓰신 이유가 있을까요..??
제가 알기론 해당 타입이 Double을 담는 타입이라 Int에 최적화되지 않은걸로 알고있어서요..! ㅎㅎㅎ

Karatuss replied:
요것도 인터넷 소스 참고하면서 필터링이 덜 된 결과입니다 ㅎㅎ. 그리고 DB마다 설계 기법이 다르다는 것을 이제 알았으니 다음부터는 DB까지 고려한 애노테이션을 삽입해보겠습니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/user/domain/User.kt:
```diff
@@ -0,0 +1,51 @@
+
+    fun withdrawMoney(withdrawMoney: Int): Int {
+        require(withdrawMoney > 0) { "출금할 값을 양수로 입력해주세요." }
+        require(withdrawMoney <= money) { "${this.id}: 현재 소지한 금액($money${LottoPrice.UNIT})보다 적은 금액을 입력해주세요." }
```
**보다 적은 금액**이 아니라
**보다 적거나** _같은_ **금액**이 맞을것 같아요~!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/user/service/UserService.kt:
```diff
@@ -0,0 +1,46 @@
+) {
+    // TODO: 나중엔 밖에서 Admin "만들지도, 접근하지도 못하도록" 인증 시스템 도입하기
+    @Transactional
+    fun createUser(request: UserCreateRequest): Long {
```
오호 요 메소드는 DB에 접속하는 코드가 단 한줄인데 트랜잭션의 범위가 필요할까요??

Karatuss replied:
더티 체킹을 위해, 혹은 예외 발생시 롤백을 하기 위해서는 선언해주는 것이 필수 아닌가요?

konempty replied:
네넵 더티체킹도 맞고 예외발생시 롤백에 쓰는것도 맞는데
해당 코드는 더티체킹도 안하고 예외가 발생할 포인트가 없어보여서 질문드려봤습니다!
만약 save에서 예외 발생하는걸 고려한 설계라면 그 경우 예외가 발생해도 롤백할데이터가 없어서 무시해도 될것 같구요!

Karatuss replied:
아하 그 포인트를 놓치고 있었네요! 해당 부분은 지우도록 하겠습니다.

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/user/service/UserService.kt:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @Transactional(readOnly = true)
+    fun getUserInformation(userId: Long): UserResponse {
```
요기도 마찬가지네요..!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/winninglotto/domain/WinningLotto.kt:
```diff
@@ -0,0 +1,36 @@
+
+@Table(
+    indexes = [
+        Index(name = "lotto_id_index", columnList = "lotto_id"),
```
아하 FK는 보통 명시적으로 인덱스라고 써두지 않아도 좋아요..!
대신 Oracle DB는 제외구요 ㅋㅋㅋ..
오라클 디비는 FK도 인덱스가 필요해요

Karatuss replied:
```suggestion
        Index(columnList = "lotto_id"),
```
그럼 이렇게 잡는게 오라클 디비 제외 일반적인 형태라고 할 수 있나요? 이러면 랜덤 스트링 생성돼서 알아보기 힘든거 아닌가요?

> 지금 보니까 DDL로는 반영하셨을지는 모르겠지만 요 엔티티 인덱스가 안잡혀있네요 ㅋㅋㅋ.. 명시적으로 어떤 인덱스가 잡혀있는지 적어두면 협업에 도움이될것 같아요~!

이전에 말씀하신 이걸 어떻게 반영해야할까요?

konempty replied:
보통 DB가 알아서 걸어주는 FK인덱스를 제외한 개발자들이 명시적으로 넣어주는 인덱스를 적어둡니다~!
따라서 FK의 인덱스는 잘 안적어주는걸 선호해요!
오라클은 FK인덱스도 적어주는게 좋구요 ㅋㅋㅋ

Karatuss replied:
아하 그럼 FK를 제외한 인덱스의 경우에만 적어주는 것이 관례라는 것이군요!
오늘도 하나 알아가네요 ㅎㅎ

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/winninglotto/repository/WinningLottoRepository.kt:
```diff
@@ -0,0 +1,10 @@
+
+@Repository
+interface WinningLottoRepository : JpaRepository<WinningLotto, Long> {
+    fun findByRound(round: Int): WinningLotto? // ?: throw NoSuchElementException("$round 라운드에 해당하는 당첨 번호가 존재하지 않습니다.")
```
헉 ㅋㅋㅋ 임시 주석같아보이는데 필요없는 코드라면 지우는게 나을것 같아요!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/winninglotto/service/WinningLottoService.kt:
```diff
@@ -0,0 +1,115 @@
+            require(round in 1..<winningLottoInformation.round) { "존재하지 않는 라운드는 조회할 수 없습니다." }
+        } else if (userRole == UserRole.ROLE_ADMIN) {
+            require(round in 1..winningLottoInformation.round) { "존재하지 않는 라운드는 조회할 수 없습니다." }
+        }
```
요 코드는 아래와 같이 고쳐도 동일한 작동을 하지 않을까요?? ㅋㅋㅋ
```kotlin
        if (userRole == UserRole.ROLE_USER) {
            require(round != winningLottoInformation.round) { "현재 진행중인 라운드는 조회할 수 없습니다." }
        } 
        require(round in 1..winningLottoInformation.round) { "존재하지 않는 라운드는 조회할 수 없습니다." }
```

Karatuss replied:
고쳐야지 했는데 결국 안 고친 과거의 신현민이 잘못했네요..ㅠ
수정했습니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/global/dto/SliceResponse.kt:
```diff
@@ -0,0 +1,15 @@
+    val hasNext: Boolean,
+    val numberOfElements: Int,
+    val content: List<T>,
+) {
```
엇 요 클래스 아무곳에서도 안쓰이는것 같은데 지워도 괜찮지 않을까요..??

Karatuss replied:
나중에 쓸까 남겨놓긴 했는데 쓸 때 다시 생성토록 하겠습니다 ㅎㅎ

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/global/exception/GlobalExceptionHandler.kt:
```diff
@@ -0,0 +1,145 @@
+    ): ResponseEntity<Any>? {
+        logger.error("[MethodArgumentNotValidException], ex")
+        val bindingResult = ex.bindingResult
+        val stringBuilder = StringBuilder()
```
오호 그냥 String을 써서 append를 하는게 아니라 명시적으로 stringBuilder를 쓰는 이유가 있을까요??

혹시
https://howtodoinjava.com/java/string/string-concatenation-performance/#2-compiler-optimization-for-string-concatenation
요 이슈 때문이라면 java 9부터 알아서 String append를 효율적인 코드로 동적 바인딩 하는 기능이 생겨서 
괜찮지 않을까 싶었어요!

Karatuss replied:
아하 고쳤습니다!

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/global/exception/GlobalExceptionHandler.kt:
```diff
@@ -0,0 +1,145 @@
+            stringBuilder.append(fieldError.field).append(":")
+            stringBuilder.append(fieldError.defaultMessage)
+            stringBuilder.append(", ")
+        }
```
에러는 어떤 방식으로든 잘 남기기만 하면 되니까 요것도 괜찮다고 생각하구요 
다른 에러가 발생햇을때도 에러를 좀더 트래킹하기 좋게 한다면 어떨까 싶어요!

그리고 한가지 알려드리고 싶은게 있는데 
나중에 외부의 로그 시스템(logstatsh, fluentd)같은걸로 중앙에 적체하는 방식을 쓰게되면
그땐 json형식으로 에러를 기록하는게 파싱하는데 편하다고 알고 계시면 도움이 될것 같아요!

Karatuss replied:
추후 중앙 로그 시스템을 구현해서 적용해보겠습니다!

konempty commented on src/test/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/repository/LottoRepositoryTest.kt:
```diff
@@ -0,0 +1,88 @@
+import org.springframework.transaction.annotation.Transactional
+
+// @RepositoryTest
+@IntegrationTest
```
요 테스트는 레포지토리만 테스트하는데 인테그레이션 테스트로 선언하는게 맞을까요..??

Karatuss replied:
shema.sql, data.sql을 아직 안 만들어서 그랬습니다 ㅎㅎ..
만든 뒤 레포지토리 테스트로 전향할 계획입니다!

konempty commented on src/test/kotlin/com/hm/hyeonminshinlottospring/support/config/KotestConfig.kt:
```diff
@@ -0,0 +1,30 @@
+ * ## 주의사항
+ *
+ * 위 내용을 이해했다면 해당 테스트 내에서 사용할 각 전역 변수들의 초기화를 롤백하기 위해서는
+ * `val`을 통해 선언할게 아니라 `lateinit var`로 설정한 뒤 `beforeEach` 함수 내에서 초기화해주어야 한다!
```
헐 ㄷㄷㄷ kotest에 이런 문제점이 있었나요..??
처음 들어보네요 회사가면 저도 확인해 봐야겠어요!

Karatuss replied:
케바케겠지만, 저 같은 경우에는 User의 userName이 unique 속성인데 이게 자꾸 제약조건에 중복된다고 걸려서 고쳤습니다.

konempty commented on src/test/kotlin/com/hm/hyeonminshinlottospring/support/converter/BooleanAttributeConverter.kt:
```diff
@@ -0,0 +1,16 @@
+import java.math.BigDecimal
+
+@Converter(autoApply = true)
+class BooleanAttributeConverter : AttributeConverter<Boolean, BigDecimal> {
```
엇 요걸 쓰신이유가 있으실가요..??

Karatuss replied:
추후에 Entity 안에 boolean 속성을 가진 프로퍼티를 대비해서 만들어뒀습니다.

konempty replied:
ㅋㅋㅋ... 요것도 오라클DB를 쓸때만 필요한 설정인걸로 알고있어요
Mysql이나 Postgresql은 Bit라는 자료형이 있어서 Bit(1) 혹은 Postgresql은 Bool로도 선언하는데
Oracle은 Bit라는 자료형이 없어서 Number(1)을 사용해서 저장 했던것 같아요
그런데 또 H2 DB를 Oracle호환모드로 키고 boolean을 Number로 저장하라고 하면 이해를 못해서 그걸 풀어내기 위해
요 코드가 필요한걸로 알고있어요

따라서 Mysql 호환 모드로 킨다면 알아서 H2DB가 알아서 잘 해줄거라고 기대되네요 ㅋㅋㅋ

Karatuss replied:
삭제 처리했습니다!

konempty reviewed (changes requested):
아핫 ㅋㅋㅋ 이번에도 리뷰에 집중해서 20개를 넘겼군요
이제 거의 다온것 같아요 조금만더 수정 부탁드릴게요!!

Karatuss committed:
docs: API 문서 추가

Karatuss committed:
docs: asciidoc 추가

- index.adoc
- lotto.adoc
- user.adoc
- winningLotto.adoc

Karatuss committed:
chore: scripts/ ignore

Karatuss committed:
build: DB 의존성 추가, 테스크 순서 지정, plain-jar 억제

Karatuss committed:
build: Docker 로컬 빌드용 스크립트 추가

Karatuss committed:
build: Docker 로컬 빌드 캐시용 스크립트 추가

Karatuss committed:
docs: 포멧 리팩토링, 코멘트 및 링크 추가

링크 목록
- Rest API Documentation
- Wiki

Karatuss committed:
refactor: 네이밍 복수형으로 변경

Karatuss committed:
refactor: 네이밍 변경 반영

Karatuss committed:
refactor: 네이밍 변경 반영, 생성 전략 롤백

Karatuss committed:
refactor: 팩토리 메서드 삭제

Karatuss committed:
feat: 마지막 라운드에 이어서 실행하는 기능 추가

Karatuss committed:
feat: pk 생성 전략 롤백

Karatuss committed:
feat: pk 생성 전략 롤백 및 모호한 표현 수정

Karatuss committed:
feat: add,withdraw -> request 파라미터 변경

Karatuss committed:
refactor: 불필요한 주석 삭제

Karatuss committed:
refactor: 로또 넘버 네이밍 변경

Karatuss committed:
feat: coroutine 도입

TODO: 추후 runBlocking 제거하기

Karatuss committed:
feat: Virtual Thread 도입

Karatuss committed:
delete: 쓰지 않는 response 클래스

Karatuss committed:
sql: flyway용 sql 초기 데이터

Karatuss committed:
env: flyway 설정 추가

Karatuss committed:
env: flyway 추가 및 DB 설정 변경

Karatuss committed:
refactor: 디테일한 에러 출력 메시지

Karatuss committed:
test: coroutine 반영 및 Rest Docs 생성 문제 해결

Karatuss committed:
test: 라운드 조회 검사 테스트 추가

Karatuss committed:
test: coroutine 설정 추가

Karatuss committed:
test: ktlint 포메팅

Karatuss committed:
refactor: todo 주석 삭제

Karatuss committed:
refactor: 로또 번호 네이밍 변경 반영

Karatuss committed:
refactor: 유저 돈 request fixture 추가

Karatuss committed:
refactor: 인자 할당 변수 지정

Karatuss committed:
delete: 통합 테스트 삭제

빌드시 발생하는 에러로 인해 삭제

Karatuss committed:
sql: h2 DB 초기화용 스키마

Karatuss committed:
docs: rest api documentation 파일

Karatuss commented:
기본적인 구현들은 다 끝낸 것 같은데 이정도면 MVP 모델이라고 할 수 있을까요? ㅋㅋㅎ
이번에도 리뷰 잘 부탁드립니다! :)

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/domain/Lotto.kt:
```diff
@@ -0,0 +1,55 @@
+import jakarta.persistence.ManyToOne
+import jakarta.persistence.Table
+
+// TODO: 최적화 솔루션) numbers를 bitset으로 다루기!!
```
오호 요건 생각도 못해본 방법인데 너무 좋습니다~! 👍 👍

konempty commented on src/main/kotlin/com/hm/hyeonminshinlottospring/domain/lotto/service/LottoService.kt:
```diff
@@ -0,0 +1,183 @@
+    ): LottoCreateResponse = runBlocking {
+        val userId = request.userId
+        val user = userRepository.findByUserId(userId)
+        check(request.insertedMoney in 0..user.money) { "${request.userId}: 현재 소지한 금액(${user.money}${LottoPrice.UNIT})보다 적은 금액을 입력해주세요." }
```
```kotlin
check(request.insertedMoney <= user.money) 
```
요런식으로 작성하는게 더 직관적이지 않을가요??

Karatuss replied:
여기서 궁금한게 있는데요, 0의 범위를 넣어준 이유가 혹시라도 컨트롤러단에서 넘어온 request가 해당 부분을 positive로 검사하지 않는다면 문제가 생기지 않을까 싶어서였습니다. 어떻게 보면 이중 체크로도 볼 수 있겠는데, 이 경우엔 어떤 식으로 처리하는게 좋을까요?

konempty replied:
Validator가 못미더웠다면 어플리케이션에서만 확인하는게 낫지않을까요..??
아니면 Validator를 통과한 이상 그냥 믿고 간다라는 선택지 밖에 없겠네요 ㅋㅋㅋㅋ

konempty commented on src/main/resources/static/docs/index.html:
```diff
@@ -0,0 +1,3158 @@
+}
+</script>
+</body>
+</html>
```
요 파일이 asciidoc에서 생성하는 파일이 아닐까 싶은데 그렇다면  gitIgnore에 넣어둬서 깃 커밋에 올라오지 않도록 하는게 좋을것 같아요!

Karatuss replied:
추가했습니다!

konempty reviewed (changes requested):
이제 진짜 MVP는 만족하는 서비스가 된것 같아서 좋네요! ㅋㅋㅋㅋ
아마 이게 찐 마지막 리뷰가 될것 같구요
별건 없고 이전에 리뷰드린 내용에 대한 현민님의 답변에 대한 재질문이 대부분이예요 ㅋㅋㅋ
진짜 쪼금만더 확인 부탁드립니다!!

Karatuss committed:
chore: ascii docs 생성 파일 추가

Karatuss committed:
refactor: 불필요한 Transactional 애노테이션 제거

Karatuss committed:
remove: Oracle DB용 컨버터 파일 삭제

Karatuss commented:
질문 하나 남겨두었고 MVP 기준으로는 해당 PR이 끝이 아닐까 싶습니다.
추후 나머지 기능을 추가해볼 예정이고 만약 그렇게 한다면 리뷰 요청 드리고 가능하시다면 봐주시면 감사하겠습니다 😄

Karatuss commented:
@laggu

konempty reviewed (approved):
크 정말 고생많으셨어요 👍 👍 
conversation이 128개라니 제 개인 신기록을 넘었네요 ㅋㅋㅋㅋ
추가로 리뷰 받고 싶은게 있다면 다음에도 리뷰를 주셔도 좋은데
제가 바쁘면 리뷰를 안달아줄수도 있다는걸 명심해주시면 될것같습니다~!

마지막으로 한가지 해소되지 않은 궁금증을 언급하자면
Pagnation vs Slice중에 Slice가 나쁘다, Paginataion을 써라 라는 의도가 아니고
둘줄 뭘쓰든 이상하지 않으며 갠취니까 괜춘하다고 생각합니다
대신 제가 계속 질문을 드린 이유는 Slice를 쓴게 독특해서 질문드렸고
요 이유는 Pagination이 DB 데이터를 전부 가져온다고 알고계셔서 였던거 맞을까요??
호옥시 다른 이유도 있었다면 간단하게 머지전에 코멘트만 적어주시면 궁금증이 해결될것 같아요!

다시 한번 진짜 고생많으셨습니다 💯

Karatuss commented:
총 데이터 개수가 필요하면 page, 그렇지 않으면 slice를 쓰면 좋다는 글을 읽은 적이 있습니다. 그래서 딱히 개수를 알 필요가 없고 무한 스크롤을 구현하면 좋겠다는 생각에 slice가 더 적합하다는 생각에 page 말고 slice를 사용했습니다.

리뷰는 당연히 한빈님이 시간이 있으실 때 봐주시는 것만으로도 감사드리죠!
다시 한번 이번 스터디를 진행해주셔서 정말 감사드리고 많은 도움 받고 갑니다 ㅎㅎ

Karatuss committed:
refactor: validator 믿기

konempty commented:
@TwoPair 크 Slice 사용사례를 정확히 알고 계시고 의도도 훌륭하네요 너무 좋습니다 👍 👍
